test/e2e/storage/testsuites/volumemode.go

```diff
@@ -104,8 +100,14 @@ func (t *volumeModeTestSuite) defineTests(driver TestDriver, pattern testpattern
 			volumeNodeAffinity *v1.VolumeNodeAffinity
 		)
 
+		l.resource = &genericVolumeTestResource{
+			driver:  driver,
+			config:  l.config,
+			pattern: pattern,
+		}
+
 		// Create volume for pre-provisioned volume tests
-		l.volume = CreateVolume(driver, l.config, pattern.VolType)
+		l.resource.volume = CreateVolume(driver, l.config, pattern.VolType)
```

Comment:
As @msau42 had pointed out, it is possible to embed `genericVolumeTestResource` directly in `local`. Then all of these `l.resource.` variable accesses can be replaced with  `l.` (as before), and the diff becomes much smaller. However, I personally find the code easier to read the way you wrote it, so I'm in favor of merging it as proposed after squashing.

FWIW, here's a patch on top  of your PR which demonstrates the other approach:
```
diff --git a/test/e2e/storage/testsuites/volumemode.go b/test/e2e/storage/testsuites/volumemode.go
index d4d12b6094..760ab0a830 100644
--- a/test/e2e/storage/testsuites/volumemode.go
+++ b/test/e2e/storage/testsuites/volumemode.go
@@ -66,9 +66,9 @@ func (t *volumeModeTestSuite) defineTests(driver TestDriver, pattern testpattern
 		config      *PerTestConfig
 		testCleanup func()
 
-		cs       clientset.Interface
-		ns       *v1.Namespace
-		resource *genericVolumeTestResource
+		cs clientset.Interface
+		ns *v1.Namespace
+		genericVolumeTestResource
 	}
 	var (
 		dInfo = driver.GetDriverInfo()
@@ -100,14 +100,14 @@ func (t *volumeModeTestSuite) defineTests(driver TestDriver, pattern testpattern
 			volumeNodeAffinity *v1.VolumeNodeAffinity
 		)
 
-		l.resource = &genericVolumeTestResource{
+		l.genericVolumeTestResource = genericVolumeTestResource{
 			driver:  driver,
 			config:  l.config,
 			pattern: pattern,
 		}
 
 		// Create volume for pre-provisioned volume tests
-		l.resource.volume = CreateVolume(driver, l.config, pattern.VolType)
+		l.volume = CreateVolume(driver, l.config, pattern.VolType)
 
 		switch pattern.VolType {
 		case testpatterns.PreprovisionedPV:
@@ -117,28 +117,28 @@ func (t *volumeModeTestSuite) defineTests(driver TestDriver, pattern testpattern
 				scName = fmt.Sprintf("%s-%s-sc-for-file", l.ns.Name, dInfo.Name)
 			}
 			if pDriver, ok := driver.(PreprovisionedPVTestDriver); ok {
-				pvSource, volumeNodeAffinity = pDriver.GetPersistentVolumeSource(false, fsType, l.resource.volume)
+				pvSource, volumeNodeAffinity = pDriver.GetPersistentVolumeSource(false, fsType, l.volume)
 				if pvSource == nil {
 					framework.Skipf("Driver %q does not define PersistentVolumeSource - skipping", dInfo.Name)
 				}
 
 				storageClass, pvConfig, pvcConfig := generateConfigsForPreprovisionedPVTest(scName, volBindMode, pattern.VolMode, *pvSource, volumeNodeAffinity)
-				l.resource.sc = storageClass
-				l.resource.pv = framework.MakePersistentVolume(pvConfig)
-				l.resource.pvc = framework.MakePersistentVolumeClaim(pvcConfig, l.ns.Name)
+				l.sc = storageClass
+				l.pv = framework.MakePersistentVolume(pvConfig)
+				l.pvc = framework.MakePersistentVolumeClaim(pvcConfig, l.ns.Name)
 			}
 		case testpatterns.DynamicPV:
 			if dDriver, ok := driver.(DynamicPVTestDriver); ok {
-				l.resource.sc = dDriver.GetDynamicProvisionStorageClass(l.config, fsType)
-				if l.resource.sc == nil {
+				l.sc = dDriver.GetDynamicProvisionStorageClass(l.config, fsType)
+				if l.sc == nil {
 					framework.Skipf("Driver %q does not define Dynamic Provision StorageClass - skipping", dInfo.Name)
 				}
-				l.resource.sc.VolumeBindingMode = &volBindMode
+				l.sc.VolumeBindingMode = &volBindMode
 
 				claimSize := dDriver.GetClaimSize()
-				l.resource.pvc = getClaim(claimSize, l.ns.Name)
-				l.resource.pvc.Spec.StorageClassName = &l.resource.sc.Name
-				l.resource.pvc.Spec.VolumeMode = &pattern.VolMode
+				l.pvc = getClaim(claimSize, l.ns.Name)
+				l.pvc.Spec.StorageClassName = &l.sc.Name
+				l.pvc.Spec.VolumeMode = &pattern.VolMode
 			}
 		default:
 			framework.Failf("Volume mode test doesn't support: %s", pattern.VolType)
@@ -146,10 +146,7 @@ func (t *volumeModeTestSuite) defineTests(driver TestDriver, pattern testpattern
 	}
 
 	cleanup := func() {
-		if l.resource != nil {
-			l.resource.cleanupResource()
-			l.resource = nil
-		}
+		l.cleanupResource()
 
 		if l.testCleanup != nil {
 			l.testCleanup()
@@ -169,22 +166,22 @@ func (t *volumeModeTestSuite) defineTests(driver TestDriver, pattern testpattern
 				var err error
 
 				By("Creating sc")
-				l.resource.sc, err = l.cs.StorageV1().StorageClasses().Create(l.resource.sc)
+				l.sc, err = l.cs.StorageV1().StorageClasses().Create(l.sc)
 				Expect(err).NotTo(HaveOccurred())
 
 				By("Creating pv and pvc")
-				l.resource.pv, err = l.cs.CoreV1().PersistentVolumes().Create(l.resource.pv)
+				l.pv, err = l.cs.CoreV1().PersistentVolumes().Create(l.pv)
 				Expect(err).NotTo(HaveOccurred())
 
 				// Prebind pv
-				l.resource.pvc.Spec.VolumeName = l.resource.pv.Name
-				l.resource.pvc, err = l.cs.CoreV1().PersistentVolumeClaims(l.ns.Name).Create(l.resource.pvc)
+				l.pvc.Spec.VolumeName = l.pv.Name
+				l.pvc, err = l.cs.CoreV1().PersistentVolumeClaims(l.ns.Name).Create(l.pvc)
 				Expect(err).NotTo(HaveOccurred())
 
-				framework.ExpectNoError(framework.WaitOnPVandPVC(l.cs, l.ns.Name, l.resource.pv, l.resource.pvc))
+				framework.ExpectNoError(framework.WaitOnPVandPVC(l.cs, l.ns.Name, l.pv, l.pvc))
 
 				By("Creating pod")
-				pod, err := framework.CreateSecPodWithNodeName(l.cs, l.ns.Name, []*v1.PersistentVolumeClaim{l.resource.pvc},
+				pod, err := framework.CreateSecPodWithNodeName(l.cs, l.ns.Name, []*v1.PersistentVolumeClaim{l.pvc},
 					false, "", false, false, framework.SELinuxLabel,
 					nil, l.config.ClientNodeName, framework.PodStartTimeout)
 				defer func() {
@@ -200,22 +197,22 @@ func (t *volumeModeTestSuite) defineTests(driver TestDriver, pattern testpattern
 				var err error
 
 				By("Creating sc")
-				l.resource.sc, err = l.cs.StorageV1().StorageClasses().Create(l.resource.sc)
+				l.sc, err = l.cs.StorageV1().StorageClasses().Create(l.sc)
 				Expect(err).NotTo(HaveOccurred())
 
 				By("Creating pv and pvc")
-				l.resource.pv, err = l.cs.CoreV1().PersistentVolumes().Create(l.resource.pv)
+				l.pv, err = l.cs.CoreV1().PersistentVolumes().Create(l.pv)
 				Expect(err).NotTo(HaveOccurred())
 
 				// Prebind pv
-				l.resource.pvc.Spec.VolumeName = l.resource.pv.Name
-				l.resource.pvc, err = l.cs.CoreV1().PersistentVolumeClaims(l.ns.Name).Create(l.resource.pvc)
+				l.pvc.Spec.VolumeName = l.pv.Name
+				l.pvc, err = l.cs.CoreV1().PersistentVolumeClaims(l.ns.Name).Create(l.pvc)
 				Expect(err).NotTo(HaveOccurred())
 
-				framework.ExpectNoError(framework.WaitOnPVandPVC(l.cs, l.ns.Name, l.resource.pv, l.resource.pvc))
+				framework.ExpectNoError(framework.WaitOnPVandPVC(l.cs, l.ns.Name, l.pv, l.pvc))
 
 				By("Creating pod")
-				pod, err := framework.CreateSecPodWithNodeName(l.cs, l.ns.Name, []*v1.PersistentVolumeClaim{l.resource.pvc},
+				pod, err := framework.CreateSecPodWithNodeName(l.cs, l.ns.Name, []*v1.PersistentVolumeClaim{l.pvc},
 					false, "", false, false, framework.SELinuxLabel,
 					nil, l.config.ClientNodeName, framework.PodStartTimeout)
 				defer func() {
@@ -240,14 +237,14 @@ func (t *volumeModeTestSuite) defineTests(driver TestDriver, pattern testpattern
 				var err error
 
 				By("Creating sc")
-				l.resource.sc, err = l.cs.StorageV1().StorageClasses().Create(l.resource.sc)
+				l.sc, err = l.cs.StorageV1().StorageClasses().Create(l.sc)
 				Expect(err).NotTo(HaveOccurred())
 
 				By("Creating pv and pvc")
-				l.resource.pvc, err = l.cs.CoreV1().PersistentVolumeClaims(l.ns.Name).Create(l.resource.pvc)
+				l.pvc, err = l.cs.CoreV1().PersistentVolumeClaims(l.ns.Name).Create(l.pvc)
 				Expect(err).NotTo(HaveOccurred())
 
-				err = framework.WaitForPersistentVolumeClaimPhase(v1.ClaimBound, l.cs, l.resource.pvc.Namespace, l.resource.pvc.Name, framework.Poll, framework.ClaimProvisionTimeout)
+				err = framework.WaitForPersistentVolumeClaimPhase(v1.ClaimBound, l.cs, l.pvc.Namespace, l.pvc.Name, framework.Poll, framework.ClaimProvisionTimeout)
 				Expect(err).To(HaveOccurred())
 			})
 		} else {
@@ -258,24 +255,24 @@ func (t *volumeModeTestSuite) defineTests(driver TestDriver, pattern testpattern
 				var err error
 
 				By("Creating sc")
-				l.resource.sc, err = l.cs.StorageV1().StorageClasses().Create(l.resource.sc)
+				l.sc, err = l.cs.StorageV1().StorageClasses().Create(l.sc)
 				Expect(err).NotTo(HaveOccurred())
 
 				By("Creating pv and pvc")
-				l.resource.pvc, err = l.cs.CoreV1().PersistentVolumeClaims(l.ns.Name).Create(l.resource.pvc)
+				l.pvc, err = l.cs.CoreV1().PersistentVolumeClaims(l.ns.Name).Create(l.pvc)
 				Expect(err).NotTo(HaveOccurred())
 
-				err = framework.WaitForPersistentVolumeClaimPhase(v1.ClaimBound, l.cs, l.resource.pvc.Namespace, l.resource.pvc.Name, framework.Poll, framework.ClaimProvisionTimeout)
+				err = framework.WaitForPersistentVolumeClaimPhase(v1.ClaimBound, l.cs, l.pvc.Namespace, l.pvc.Name, framework.Poll, framework.ClaimProvisionTimeout)
 				Expect(err).NotTo(HaveOccurred())
 
-				l.resource.pvc, err = l.cs.CoreV1().PersistentVolumeClaims(l.resource.pvc.Namespace).Get(l.resource.pvc.Name, metav1.GetOptions{})
+				l.pvc, err = l.cs.CoreV1().PersistentVolumeClaims(l.pvc.Namespace).Get(l.pvc.Name, metav1.GetOptions{})
 				Expect(err).NotTo(HaveOccurred())
 
-				l.resource.pv, err = l.cs.CoreV1().PersistentVolumes().Get(l.resource.pvc.Spec.VolumeName, metav1.GetOptions{})
+				l.pv, err = l.cs.CoreV1().PersistentVolumes().Get(l.pvc.Spec.VolumeName, metav1.GetOptions{})
 				Expect(err).NotTo(HaveOccurred())
 
 				By("Creating pod")
-				pod, err := framework.CreateSecPodWithNodeName(l.cs, l.ns.Name, []*v1.PersistentVolumeClaim{l.resource.pvc},
+				pod, err := framework.CreateSecPodWithNodeName(l.cs, l.ns.Name, []*v1.PersistentVolumeClaim{l.pvc},
 					false, "", false, false, framework.SELinuxLabel,
 					nil, l.config.ClientNodeName, framework.PodStartTimeout)
 				defer func() {
```